### PR TITLE
Ethereum transfer: add support for ERC1155/safeTransferFrom

### LINKED ISSUE #1295: Add a new type of EthereumTransactionType for supporting ERC1155 transfer
Similar to `EthereumTransactionType.erc721Transfer` needs to add ability to transfer ERC1155. 
ERC1155 uses `safeTransferFrom` function 

```
function safeTransferFrom(address _from, address _to, uint256 _tokenId, uint256 _amount, bytes _data)
```

Also it would be useful  that `EthereumTransactionType` returns the `data` value of the function input. It will allow to remove abi encoders from the app.

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Ethereum transfer: add support for ERC1155/safeTransferFrom .
Fixes #1295 .

## Testing instructions

Ethereum unit tests.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/proto/Ethereum.proto [hunk 1]
```diff
@@ -40,6 +40,20 @@ message Transaction {
...
+        string from = 1;
+
+        string to = 2;
+
+        // ID of the token (256-bit number)
+        bytes token_id = 3;
+
+        bytes value = 4;
+
+        bytes calldata = 5;
```
Comment (hewigovens) [inline src/proto/Ethereum.proto:54]:
```suggestion
        bytes data = 5;
```

**Review by hewigovens — APPROVED** [review]
looks good, better to add one swift or kotlin test just in case

File: src/Ethereum/Transaction.cpp [hunk 2]
```diff
@@ -58,3 +63,16 @@ Data Transaction::buildERC721TransferFromCall(const Data& from, const Data& to,
     func.encode(payload);
     return payload;
 }
+
+Data Transaction::buildERC1155TransferFromCall(const Data& from, const Data& to, uint256_t tokenId, uint256_t value, const Data& data) {
```
Comment (vikmeup) [inline src/Ethereum/Transaction.cpp:67]:
```
@param _from    Source address
        @param _to      Target address
        @param _ids     IDs of each token type (order and length must match _values array)
        @param _values  Transfer amounts per token type (order and length must match _ids array)
        @param _data    Additional data with no specified format, MUST be sent unaltered in call to the `ERC1155TokenReceiver` hook(s) on `_to`
    */
```

Should we support multiple `Ids` and `values` and values as per spec: https://eips.ethereum.org/EIPS/eip-1155

File: src/proto/Ethereum.proto [hunk 3]
```diff
@@ -40,6 +40,20 @@ message Transaction {
...
+    // ERC1155 NFT transfer transaction
+    message ERC1155Transfer {
+        string from = 1;
+
+        string to = 2;
+
+        // ID of the token (256-bit number)
+        bytes token_id = 3;
+
+        bytes value = 4;
```
Comment (vikmeup) [inline src/proto/Ethereum.proto:52]:
```suggestion
        // The amount of tokens being transferred
        bytes value = 4;
```

File: src/Ethereum/Transaction.cpp (hunk 2 above)
Comment (hewigovens) [reply src/Ethereum/Transaction.cpp:67]: that's another call `safeBatchTransferFrom`, we can add this later if we need to support transfer multiple collectibles in the app

**Review by vikmeup — APPROVED** [review]

**Pull request merged**